### DISCUSSION TITLE
troubleshooting oh-my-posh glyphs / terminal-icons

### DISCUSSION THREAD

**Discussion by noelle-gray** [JanDeDobbeleer/oh-my-posh#3790] [Troubleshoot] [answered]
Hello,

I'm rather embarrassed to bring this up but I've been troubleshooting for two days now. I cannot get many glyphs to display properly on Terminal.

I should start out by saying I'm native and fluent with Bash/Unix environments. I like to use oh-my-posh as a way of prettying up whenever I remote into a server or need work done on Windows while gaming. 

I've been trying to read through every FAQ and documentation and still can't get this fixed, so I'm here.

I will **fully** admit that my install process has been confusing. I feel like the way to fix this is the equivalent of git re-basing my entire Windows. I just want a pretty terminal, dangit.

## Issue: Glyphs not mapping correctly to oh-my-posh installation.

![Screenshot 2023-05-03 230928](https://user-images.githubusercontent.com/121739115/236070815-a85a1a0b-04bd-44d6-914e-134cf952a50a.png)

## Debug Details

```

California@MyPc  阮  Olivier Libaux - No One Knows   ~  ﲵ pwsh   
Version: 15.4.1

Shell: powershell (5.1.22621.963)

Segments:

ConsoleTitle(false) -   0 ms -
session(true)      -   1 ms -  California@MyPC
spotify(true)      -   4 ms -  阮  Bedřich Smetana - Má vlast (My Country): No. 2, Vltava [Moldau]
path(true)         -   0 ms -   ~
git(false)         -   2 ms -
node(false)        -   0 ms -
shell(true)        -   0 ms -  ﲵ powershell
root(false)        -   0 ms -
exit(true)         -   0 ms -   

Run duration: 17.5341ms
```

## Config details:

```
Cache path: C:\Users\California\AppData\Local\oh-my-posh
Config path: C:\Users\California\AppData\Local\Programs\oh-my-posh\themes\plague.omp.json
```

## Logs:
```
[DEBUG] 21:17:49.650 shell.go:Getenv:394 → NO DATA
[TRACE] 21:17:49.650 shell.go:Getenv(TERM_PROGRAM) - 0s
[DEBUG] 21:17:49.650 shell.go:Getenv:394 → C:\Users\California\AppData\Local
[TRACE] 21:17:49.650 shell.go:Getenv(LOCALAPPDATA) - 0s
[TRACE] 21:17:49.650 shell_windows.go:CachePath() - 0s
[DEBUG] 21:17:49.651 shell.go:Getenv:394 → C:\Users\California\AppData\Local\Programs\oh-my-posh\themes\plague.omp.json
[TRACE] 21:17:49.651 shell.go:Getenv(POSH_THEME) - 0s
[DEBUG] 21:17:49.657 shell.go:Shell:641 → process name: powershell.exe
[TRACE] 21:17:49.657 shell.go:Shell() - 6.1981ms
[TRACE] 21:17:49.657 shell.go:resolveConfigPath() - 6.1981ms
[TRACE] 21:17:49.657 shell.go:Init() - 6.8448ms
[TRACE] 21:17:49.657 shell.go:Flags() - 0s
[TRACE] 21:17:49.657 config.go:loadConfig() - 514.6µs
[TRACE] 21:17:49.657 shell.go:Flags() - 0s
[DEBUG] 21:17:49.657 shell.go:Getenv:394 → NO DATA
[TRACE] 21:17:49.657 shell.go:Getenv(OMP_CACHE_DISABLED) - 0s
[TRACE] 21:17:49.657 shell_windows.go:WindowsRegistryKeyValue(HKEY_CURRENT_USER\Software\Microsoft\Windows\DWM\ColorizationColor) - 0s
[TRACE] 21:17:49.657 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.658 shell_windows.go:WindowsRegistryKeyValue:206 → ColorizationColor(DWORD): 0xC4D0000C
[TRACE] 21:17:49.658 shell.go:Shell() - 0s
[DEBUG] 21:17:49.658 shell.go:Getenv:394 → 5.1.22621.963
[TRACE] 21:17:49.658 shell.go:Getenv(POSH_SHELL_VERSION) - 0s
[DEBUG] 21:17:49.658 debug.go:PrintDebug:22 → Segment: Title
[TRACE] 21:17:49.658 shell.go:Flags() - 0s
[DEBUG] 21:17:49.658 block.go:Debug:230 → Segment: session
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.658 shell.go:Pwd:424 → C:\Users\California
[TRACE] 21:17:49.658 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.658 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.658 shell.go:Getenv:394 → NO DATA
[TRACE] 21:17:49.658 shell.go:Getenv(SSH_CONNECTION) - 0s
[DEBUG] 21:17:49.658 shell.go:Getenv:394 → NO DATA
[TRACE] 21:17:49.658 shell.go:Getenv(SSH_CLIENT) - 0s
[TRACE] 21:17:49.658 shell_windows.go:Root() - 0s
[TRACE] 21:17:49.658 shell.go:Shell() - 0s
[TRACE] 21:17:49.658 shell.go:ErrorCode() - 0s
[TRACE] 21:17:49.658 shell_windows.go:IsWsl() - 0s
[TRACE] 21:17:49.658 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.658 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.658 shell.go:User:545 → California
[TRACE] 21:17:49.658 shell.go:User() - 0s
[DEBUG] 21:17:49.658 shell.go:Host:557 → MyPC
[TRACE] 21:17:49.658 shell.go:Host() - 0s
[TRACE] 21:17:49.658 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.658 shell.go:Getenv:394 → NO DATA
[TRACE] 21:17:49.658 shell.go:Getenv(SHLVL) - 0s
[TRACE] 21:17:49.658 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.660 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.660 shell.go:Shell() - 0s
[DEBUG] 21:17:49.660 block.go:Debug:230 → Segment: spotify
[TRACE] 21:17:49.660 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.660 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.660 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell_windows.go:QueryWindowTitles(^(Spotify.*)|(.*\s-\s.*)$) - 4.1596ms
[DEBUG] 21:17:49.664 properties.go:GetString:28 → 
[TRACE] 21:17:49.664 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.664 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.664 shell.go:Shell() - 0s
[DEBUG] 21:17:49.664 block.go:Debug:230 → Segment: path
[TRACE] 21:17:49.664 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.664 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:Pwd() - 0s
[TRACE] 21:17:49.664 shell.go:Shell() - 0s
[TRACE] 21:17:49.664 shell.go:Shell() - 0s
[TRACE] 21:17:49.664 shell.go:Flags() - 0s
[DEBUG] 21:17:49.664 properties.go:GetBool:22 → mapped_locations_enabled: true
[DEBUG] 21:17:49.664 properties.go:GetString:28 → 
[DEBUG] 21:17:49.664 properties.go:GetString:28 → 
[DEBUG] 21:17:49.664 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.664 properties.go:GetString:28 → ~
[DEBUG] 21:17:49.664 properties.go:GetKeyValueMap:46 → mapped_locations: map[]
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:GOOS() - 0s
[TRACE] 21:17:49.664 shell.go:PathSeparator() - 0s
[TRACE] 21:17:49.664 shell.go:Pwd() - 0s
[TRACE] 21:17:49.664 shell_windows.go:IsWsl() - 0s
[TRACE] 21:17:49.664 shell.go:StackCount() - 0s
[DEBUG] 21:17:49.665 win32_windows.go:isWriteable:299 → not current user or in group
[DEBUG] 21:17:49.665 win32_windows.go:isWriteable:303 → current user is member of S-1-5-32-544
[DEBUG] 21:17:49.665 win32_windows.go:isWriteable:311 ↓
    WRITE_DAC
    WRITE_OWNER
    SYNCHRONIZE
    DELETE
    READ_CONTROL
[DEBUG] 21:17:49.665 win32_windows.go:isWriteable:313 → user has write access
[TRACE] 21:17:49.665 shell_windows.go:DirIsWritable() - 517.7µs
[TRACE] 21:17:49.665 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.665 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.665 shell.go:Shell() - 0s
[DEBUG] 21:17:49.665 block.go:Debug:230 → Segment: git
[TRACE] 21:17:49.665 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.665 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.665 shell.go:GOOS() - 0s
[TRACE] 21:17:49.665 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.667 shell.go:CommandPath:596 → C:\Program Files\Git\cmd\git.exe
[TRACE] 21:17:49.667 shell.go:CommandPath(git.exe) - 2.0673ms
[TRACE] 21:17:49.667 shell.go:HasCommand(git.exe) - 2.0673ms
[TRACE] 21:17:49.667 shell.go:Pwd() - 0s
[ERROR] 21:17:49.667 shell.go:HasParentFilePath:729 → CreateFile .git: The system cannot find the file specified.
[TRACE] 21:17:49.667 shell.go:HasParentFilePath(.git) - 0s
[DEBUG] 21:17:49.667 properties.go:GetBool:22 → fetch_bare_info: false
[DEBUG] 21:17:49.667 block.go:Debug:230 → Segment: node
[TRACE] 21:17:49.667 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.667 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.667 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.667 properties.go:GetStringArray:52 → extensions: [*.js *.ts package.json .nvmrc pnpm-workspace.yaml .pnpmfile.cjs .npmrc .vue]
[DEBUG] 21:17:49.667 properties.go:GetStringArray:52 → folders: []
[DEBUG] 21:17:49.667 properties.go:GetBool:22 → home_enabled: false
[TRACE] 21:17:49.667 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.667 shell_windows.go:56 → C:\Users\California
[DEBUG] 21:17:49.667 block.go:Debug:230 → Segment: shell
[TRACE] 21:17:49.667 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.667 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.667 shell.go:GOOS() - 0s
[DEBUG] 21:17:49.667 properties.go:GetKeyValueMap:46 → mapped_shell_names: map[]
[TRACE] 21:17:49.667 shell.go:Shell() - 0s
[TRACE] 21:17:49.667 shell.go:Flags() - 0s
[TRACE] 21:17:49.667 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.667 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.667 shell.go:Shell() - 0s
[DEBUG] 21:17:49.667 block.go:Debug:230 → Segment: root
[TRACE] 21:17:49.667 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.667 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.667 shell.go:GOOS() - 0s
[TRACE] 21:17:49.668 shell_windows.go:Root() - 133.6µs
[DEBUG] 21:17:49.668 block.go:Debug:230 → Segment: exit
[TRACE] 21:17:49.668 shell.go:Pwd() - 0s
[DEBUG] 21:17:49.668 shell_windows.go:56 → C:\Users\California
[TRACE] 21:17:49.668 shell.go:GOOS() - 0s
[TRACE] 21:17:49.668 shell.go:ErrorCode() - 0s
[DEBUG] 21:17:49.668 properties.go:GetBool:22 → always_enabled: true
[TRACE] 21:17:49.668 shell.go:TemplateCache() - 0s
[TRACE] 21:17:49.668 shell.go:Shell() - 0s
[TRACE] 21:17:49.668 shell.go:TemplateCache() - 0s
[DEBUG] 21:17:49.668 shell.go:Getenv:394 → C:\Users\California\AppData\Local
[TRACE] 21:17:49.668 shell.go:Getenv(LOCALAPPDATA) - 0s
[TRACE] 21:17:49.668 shell_windows.go:CachePath() - 0s
[TRACE] 21:17:49.668 shell.go:Flags() - 0s
```

## .ps1 profile:

```
Import-Module terminal-icons
oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH/plague.omp.json" | Invoke-Expression
```

I've tried all three methods of installing oh-my-posh (winget, scoop, manual- in that order). I suspect it is something to do with this. I have removed scoop from my system entirely.

## Relevant details

Attempted fix:

- `winget` install method (exact from documentation)
- Multiple installation methods (I know, bad idea, but we're already past that)
- Reset my psmodulepaths
- Trying all methods of installing terminal-fonts
- Making an exception for OMP in Windows Security Settings
- [This ](https://github.com/devblackops/Terminal-Icons/issues/101) issue, given it's timing with my problem, so pushed that in my terminal once.
- Setting UTF-8 on .ps1 with `$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding' + [Environment]::Newline + (Get-Content -Raw $PROFILE -ErrorAction SilentlyContinue) | Set-Content -Encoding utf8 $PROFILE`

Brief notes:
- Yes, I'm using NerdFonts. It doesn't matter which NerdFont is used - none of the affected glyphs are appearing correctly - however, they are NOT the standard boxes associated with first install.

- Terminal default settings:
  - MesloLGM NerdFont (among a bunch of others I tried)
  - `AtlasEngine` enabled

**I believe the fundamental issue is that terminal-icons is not mapping correctly with oh-my-posh installation.** However, I do not know enough about the source code or its Go to debug it.

I am hoping the provided stacktrace provides a little help for my troubleshooting problems. Additionally, I have some questions.

## Detailed Questions

**Questions**:

- What is the standard way for using `terminal-icons` with oh-my-posh for the winget installation method in PowerShell? I can't seem 
- Do I need to declare `Import-Module terminal-icons` in my profile.ps1 in the current version of oh-my-posh? I haven't found this in the documentation and various tutorials. A FAQ does mention to remove `Import-Module` from the .ps1 profile here 
- Is it necessary to specifically edit the `settings.json` file itself or does changing the font in Terminal Settings do the job?
- Are there any known conflicts between install methods and PowerShell module locations involving `terminal-icons` that might alter where they are called from, such as a static reference in a previous install?

**Quick Comments**:
- It's possible I borked my oh-my-posh installation in the scoop method of installation. I don't know how, just a gut feeling.
- After the scoop method didn't work, I tried scoop + choco version of terminal-icons. A manual installation of the .nupkg was even attempted for terminal-icons.

I hope someone here is able to guide me through some of these hang-ups for why my oh-my-posh can't render glyphs correctly. I'm nearly positive it's terminal-icons module not working properly with my oh-my-posh installation. This is a detail I can't seem to find in the documentation for. 

A lot of the existing help sources give lots of configs and help for previous versions, which has been confusing to compare with existing documentation.

If I can't get this to work through the Windows terminal, I'll probably look at if I can get a ConEmu/other terminal version working. I'm concerned that if I can't get this working on the Windows terminal it'll be even worse with another emulator.

I know this is a lot of info to drop just to get glyphs working right. It might be better to post over at terminal-icons, but I want to check whether it's an issue with the OMP installation finding the terminal-icons, not the terminal-icons.

Hoping this is a simple easy fix I'm missing:)

Best, 

Gray

**Comment by noelle-gray** [comment, marked as answer]
The solution was `oh-my-posh config migrate glyphs --write`.

I think my confusion was the migration process for the Nerd Fonts had been applied but without success.

Given the several installation methods attempted with the correct fix method applied without success. I think the reason I was fooled was verifications of static refs with a cached font. 

I noticed that `oh-my-posh/src/engine/migrate_glyphs.go` references a static map to migrate glyph values. Just throwing out there that it looks like a case could be made for improving `TestGetCodePoints` from checking the length of a key-value map of a single in-house font font to checking the actual font being sent to the terminal emulator in use. It may be worth it to do a simple check and say "hey, this is a v2 font- you're gonna want to switch to a v3" through a CLI call such as `oh-my-posh validate fonts`.

I cannot confirm why the given quickfix did not work for me, but I can confirm that it was tried several times earlier today. However, my uneducated guess think that in between my first installation of oh-my-posh and various attempts at installing `nerd-fonts`, `scoop`, `winget`, and `install-module` introduced a copy of nerd fonts valid for `migrate glyphs` somewhere in the cache. 

It appears that I'm particularly unlucky in my timing to install oh-my-posh with the newest release of Nerd Fonts v3, with relevant issues https://github.com/JanDeDobbeleer/oh-my-posh/issues/3781 and [ryanoasis/nerd-fonts#365](https://github.com/ryanoasis/nerd-fonts/issues/365) popping up around the same time.

Anyway, I'm happy there's a fix and I hope everyone is entertained enough by my fumbling to not make the same mistakes.

Gray

**Comment by sebbean** [comment]
@noelle-gray think i'm having the same issue and that command didnt seem to help me even after a close and reopen of terminal (vscode hosted powershel)
just noticing that the standalone powershell works ok 

seems like vscode (cursor really) isnt configured correctly and not respecting my theme config change either

no real question - more of a papertrail - i'll report back if / when i sort it out
